Userland/Libraries/LibGfx/Painter.cpp

```diff
@@ -1248,6 +1248,9 @@ ALWAYS_INLINE static void do_draw_scaled_bitmap(Gfx::Bitmap& target, IntRect con
     case Painter::ScalingMode::BilinearBlend:
         do_draw_scaled_bitmap<has_alpha_channel, Painter::ScalingMode::BilinearBlend>(target, dst_rect, clipped_rect, source, src_rect, get_pixel, opacity);
         break;
+    case Painter::ScalingMode::None:
+        do_draw_scaled_bitmap<has_alpha_channel, Painter::ScalingMode::None>(target, dst_rect, clipped_rect, source, src_rect, get_pixel, opacity);
+        break;
```

Comment:
I'm not sure about this. The `do_draw_scaled_bitmap()` this calls doesn't have any handling for `Painter::ScalingMode::None`, so I think it will fall back to some kind of scaling, which may or may not even be valid.

I don't know what's correct here. Maybe `VERIFY()` here that the `src_rect` and `dst_rect` are the same size, and then call `blit_bitmap()` instead.

Reply:
Added some handling to `draw_scaled_bitmap` to handle these situations. What it does is create a "clipped" rect of the source to the destination dimensions, and blits that to the destination location. Obviously there's no current code that calls this in Serenity, but a minimum sample would be something like the diff below. Then you can run FlappyBug and verify that the cloud is clipped to half it's height. You can change the `0.5f` value in `Game.h` to for example `2.5f` and see that we don't scale the image to that dimension, we just show what we have in the source at it's original scale.

![image](https://user-images.githubusercontent.com/227077/199381139-0f643c2d-6939-4bdb-8d37-c1a6f577d50e.png)


```
diff --git a/Userland/Games/FlappyBug/Game.cpp b/Userland/Games/FlappyBug/Game.cpp
index e8fc4946bd..5ed660dd8d 100644
--- a/Userland/Games/FlappyBug/Game.cpp
+++ b/Userland/Games/FlappyBug/Game.cpp
@@ -63,7 +63,7 @@ void Game::paint_event(GUI::PaintEvent& event)
 
     painter.draw_tiled_bitmap(frame_inner_rect(), *m_background_bitmap);
 
-    painter.draw_scaled_bitmap(m_cloud.rect(), *m_cloud.bitmap(), m_cloud.bitmap()->rect(), 0.2f);
+    painter.draw_scaled_bitmap(m_cloud.rect(), *m_cloud.bitmap(), m_cloud.bitmap()->rect(), 0.2f, Gfx::Painter::ScalingMode::None);
 
     painter.fill_rect(enclosing_int_rect(m_obstacle.top_rect()), m_obstacle.color);
     painter.fill_rect(enclosing_int_rect(m_obstacle.bottom_rect()), m_obstacle.color);
diff --git a/Userland/Games/FlappyBug/Game.h b/Userland/Games/FlappyBug/Game.h
index 74137707b9..95684fd9d1 100644
--- a/Userland/Games/FlappyBug/Game.h
+++ b/Userland/Games/FlappyBug/Game.h
@@ -160,7 +160,7 @@ public:
 
         Gfx::IntRect rect() const
         {
-            return { (int)x - bitmap()->width(), (int)y - bitmap()->height(), bitmap()->width(), bitmap()->height() };
+            return { (int)x - bitmap()->width(), (int)y - bitmap()->height(), bitmap()->width(), (int)(bitmap()->height()  * 0.5f)};
         }
     };
 

```